### DISCUSSION TITLE
Add support for sd_bus_set_method_call_timeout()

### DISCUSSION THREAD

**Discussion by meehand** [python-sdbus/python-sdbus#53] [Ideas]
As it stands, there is no support in the library for adjusting the timeout away from the default. If support for sd_bus_set_method_call_timeout() was added, we could adjust the timeout.

**Comment by igo95862** [comment]
Hello @meehand 

Sounds pretty straight forward to add.

However, is there a particular reason you want to tweak timeout?

**Reply by meehand** [reply]
Without going into too much detail I have a dbus command interface that allows me to run a hardware feature for an arbitrary amount of time. Once it completes it responds. As it stands I can only run that HW feature for a max of 25s before I no longer receive the response.

Another option is to open up access to the timeout argument that's ultimately sent to sd_bus_call.

**Reply by igo95862** [reply]
I see.

A bit strange implementation. I would have used signals. For example, Freedestkop Notifications returns a unique id when notification is created that can be later used to monitor signals when notification gets closed or invokes action.

> Another option is to open up access to the timeout argument that's ultimately sent to sd_bus_call.

Hmm. Not sure how to implement this in current ORM. Maybe with a `with` block?

```python
with dbus_timeout(60):
    await long_method()
```

**Reply by meehand** [reply]
It originally was designed to only be run for a small amount of time, but we're taking it outside of that original design. 

I was was messing around with a fork and had it working with timeout as an additional member of DbusMethodCommon then just flowing down to the ultimate sd_bus_call. I'm sure you may object to that design though. I'm  not opposed to the with block

**Comment by igo95862** [comment]
Hmm... Reading the man page aparently there is `SYSTEMD_BUS_TIMEOUT` environment variable to control the timeout.

https://man.archlinux.org/man/sd_bus_set_method_call_timeout.3.en#DESCRIPTION

**Reply by meehand** [reply]
Yeah, it's cached though, so no changing it at runtime.

**Comment by igo95862** [comment]
@meehand I added the `SdBus.method_call_timeout_usec` property in e7f3bb115ec08790cfdba5f35bcd9178d1eecc15. Can you give it a test? (remember that it is in microseconds)

As for controlling the timeout per call I would look in to that more in the future.

**Reply by meehand** [reply]
Trying it out now

**Reply by meehand** [reply]
This worked great!

**Comment by igo95862** [comment]
I just realized that it is possible to have extra methods for D-Bus method descriptor objects. So there can be an extra method to do a call with extra timeout. For example, calling some long_method with 60 second timeout and some_argument:

```python
await  dbus_object.long_method.with_timeout(60)(some_argument)
```

I think the typing should work out.

**Reply by colazzo** [reply]
That seems to be a nice approach for setting timeout on method level, is this something you already added @igo95862  or are you planning to add this functionality in next version?

**Reply by igo95862** [reply]
I haven't added it yet. I might add it to 0.12.0 or 0.13.0. The question I have is how it should interact when calling local object? Should it not be allowed and raise `RuntimeError` as soon as called or should use asyncio `wait_for` and raise a `TimeoutError`?

**Reply by igo95862** [reply]
If it has a use I can add it in 0.12.0. It shouldn't be that difficult to implement.

**Reply by colazzo** [reply]
There is the DbusNoReplyError already, can you use this one?
I was testing the one you added on bus level by setting this property:
```
      bus: SdBus = get_default_bus()
      bus.method_call_timeout_usec = 100_000
```
By doing this I would expect to timeout directly after 100ms right?
When my service is blocked for 1 minute, the client call does not go into timeout after 100 ms, it had to wait 1 minute instead. I would expect to see the DbusNoReplyError exception directly after 100 ms.

**Reply by igo95862** [reply]
Did you set the timeout on client or server bus?

**Reply by colazzo** [reply]
On client side, after setting new_proxy or should it be set before the new_proxy call?

**Reply by igo95862** [reply]
It shouldn't matter because default timeout should be used every call. I will investigate this. (by the way you can use `asyncio.wait_for` to implement timeouts in Python)

**Reply by igo95862** [reply]
I think I found a work-around:

```python
    async def test_bus_timeout(self) -> None:
        test_object, test_object_connection = initialize_object()
        test_object_connection._dbus.attached_bus.method_call_timeout_usec = 100_000

        async def test():
            while True:
                await test_object_connection.returns_none_method()
                await sleep(0.5)

        t = get_running_loop().create_task(test())

        await test_object_connection.looong_method()  # Has await sleep(100)
```

The `await test_object_connection.looong_method()` will timeout after 0.5 seconds. Looks like the timeouts can't fire properly unless there is some other activity on the bus. Not sure if it an integration problem or not.

**Reply by igo95862** [reply]
I think the issue is I am not calling the `sd_bus_get_timeout` wait for a specific period of time. However, integrating it with the Python's asyncio loop is problematic as you can only add file descriptors to watch and not control specific timeout time the epoll will be called for. However, I have an idea that might work.

**Reply by colazzo** [reply]
My original example of a timeout of 100 ms was perhaps too extreme I must say :-)
If we manage to have min. timeout of 1 second and precision of around 0.5 seconds is also fine.

More important is being able to have the longer timeouts than the default one of 25 seconds.
Ideally being able to set per method.

**Reply by igo95862** [reply]
The longer timeout should already be allowed by setting `bus.method_call_timeout_usec` to a large value. It is just the shorter timeout might not fire on time because the timeouts are only processed when there is data to read from D-Bus connection.

**Reply by colazzo** [reply]
Hi @igo95862 I think the timeout setting is not documented yet? Could you update the doc with this feature... also explain on which side you can apply the timeout (client side, or server side...).

**Reply by igo95862** [reply]
It is undocumented for now as all low-level API because it is not finalized yet.

**Reply by igo95862** [reply]
The high level interface

```python
await  dbus_object.long_method.with_timeout(60)(some_argument)
```

can definitely work but the Jedi code auto completion does not support the ParamSpec yet. Without it there won't be function arguments help anywhere. Right now the async dbus method masquerade as an async functions.
